storybookjs/storybook
# React: Upgrade `react-docgen` to 6.0.x and improve argTypes

Extends #22324 

## What I did

**NOTE: This only applies to _VITE_ projects, not Webpack ones yet!!!**

Extends  #22324:
- [x] Update `react-docgen` to 6.0.x
- [x] Add support for fs resolver

Followup:
- [ ] Update webpack too

Also improve argType generation:
- [x] Add experimental support for converting `unions` of `literals` into `enum`, which generate proper controls
- [x] Fix table summary types for array, object, and function

## How to test in your project

To test this out in a React Vite TS project, install the canary version listed at the top of this PR into a project by manually updating its `package.json` to update all the `@storybook/*` packages from this monorepo:

```
    "@storybook/react": "0.0.0-pr-23825-sha-7875282b",
    "@storybook/react-vite": "0.0.0-pr-23825-sha-7875282b",
    ...
```

Once you've installed the packages, edit `.storybook/main.ts` to contain:
```
  typescript: {
    reactDocgen: 'react-docgen',
  },
```

The behavior should be the same as before, but running `storybook dev` should be dramatically faster than before.

If you run into issues, please comment on this PR!

## How to test in a development sandbox

1. Create a development sandbox in this branch, e.g.
```
yarn task --task sandbox --start-from auto --template react-vite/default-ts
```
2. Edit `.storybook/main.ts` to contain:
```
  typescript: {
    reactDocgen: 'react-docgen',
  },
```
3. Add a new file `starbucks-size.ts` in `src/stories` directory:
```ts
export type StarbucksSize = 'tall' | 'grande' | 'venti';
```
4. Update the CLI-generated `Button.tsx` component with a new prop:
```ts
  /**
   * How large should the button be at Starbucks?
   */
  starbucksSize?: StarbucksSize;
```
5. Observe the result in the controls panel:

<img width="612" alt="image" src="https://github.com/storybookjs/storybook/assets/488689/2b8b772d-43ef-4cdc-9d83-368b20523dfa">


## Checklist

Needs tests:

### 🦋 Canary release


This pull request has been released as version [`0.0.0-pr-23825-sha-7875282b`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-23825-sha-7875282b). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-23825-sha-7875282b`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-23825-sha-7875282b) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/fix-react-docgen-enum`](https://github.com/storybookjs/storybook/tree/shilman/fix-react-docgen-enum) |
| **Commit** | [`7875282b`](https://github.com/storybookjs/storybook/commit/7875282ba0dfebc4e5f6a59ead02e08c7e509d3a) |
| **Datetime** | Tue Aug 22 10:30:44 UTC 2023 (`1692700244`) |
| **Workflow run** | [5937622254](https://github.com/storybookjs/storybook/actions/runs/5937622254) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=23825`_
</details>
